DRIVERS-2577: add a runCommand specification (#1389)
mongodb/specifications: source/run-command/run-command.rst

META
====

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
document are to be interpreted as described in `RFC 2119
<https://www.ietf.org/rfc/rfc2119.txt>`_.

Specification
=============

-----
Terms
-----
Command
    A structure representing a BSON document that has a shape matching a supported MongoDB operation.

---------------------------
Implementation requirements
---------------------------

All drivers MAY offer the operations defined in the following sections.
This does not preclude a driver from offering more.

--------------
``runCommand``
--------------

The following represents how a runCommand API SHOULD be exposed.

.. code:: typescript

    interface Db {
      /**
       * Takes an argument representing an arbitrary BSON document and executes it against the server.
       */
      runCommand(command: BSONDocument, options: RunCommandOptions): BSONDocument;
    }

    interface RunCommandOptions {
      /**
       * An optional readPreference setting to apply to server selection logic.

>        * Also to apply to the command document as the $readPreference global command argument.

Author: Must, fixed

       *
       * @defaultValue primary
       *
       * @see https://www.mongodb.com/docs/manual/core/read-preference/
       */
      readPreference?: ReadPreference;

      /**
       * An optional explicit client session.

>        * The associated logical session id (`lsid`) will be applied to the command.

Author: Must, fixed

       *
       * @see https://www.mongodb.com/docs/manual/core/read-isolation-consistency-recency/#std-label-sessions
       */
      session?: ClientSession;

      /**
       * An optional duration the command should be permitted to run for.
       */
      timeoutMS?: number;
    }

Deviations
^^^^^^^^^^

* Drivers MAY rename their API to match the language syntax.

RunCommand implementation details
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

RunCommand provides a way to access MongoDB server commands directly without requiring a driver to implement a bespoke helper.

> The API is intended to take a document from a user, apply a number of common driver internal concerns before forwarding the command to a server.

Author: fixed

In general, a driver SHOULD avoid inspecting or modifying the user's command document in an effort to remain forward compatible with any potential server commands.

> Drivers that have historically modified user input SHOULD strive to instead clone so modifications do not impact reusing the same input command document.

Author: I think I was focusing on a use case, what matters is not to modify the user's input, I'm thinking that is a useful garuntee so that a command document could be reused by the user who made it.

OP_MSG
""""""

The ``$db`` global command argument MUST be sent on the command sent to the server and it MUST correspond to the database name the runCommand API was invoked on.

* See OP_MSG's section on `Global Command Arguments <https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#global-command-arguments>`_

ReadPreference
""""""""""""""

A driver's runCommand implementation MUST assume all commands are read operations.
To facilitate selecting the server the command should be sent to a driver's runCommand MUST accept an optional Read Preference option.

* See Server Selection's section on `Use of read preferences with commands <https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#use-of-read-preferences-with-commands>`_

If the provided Read Preference is NOT primary, the command sent MUST include the ``$readPreference`` global command argument.

* See OP_MSG's section on `Global Command Arguments <https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#global-command-arguments>`_

Driver Sessions
"""""""""""""""

> A driver's runCommand API MUST provide an optional session option to support explicit sessions and transactions.

Author: Good call, added a sentence

Every ClientSession has a corresponding Logical Session ID representing the server side session ID.

> The LSID MUST be included under ``lsid`` in the command sent to the server without modifying user input.

Author: fixed

* See Driver Sessions' section on `Sending the session ID to the server on all commands <https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#sending-the-session-id-to-the-server-on-all-commands>`_

Transactions
""""""""""""

If RunCommand is used within a Transaction the read preference MUST be sourced from the Transaction's options.

> The command sent to the server MUST include the transaction specific fields, the following is a summary:

Author: fixed

* If ``runCommand`` is executing within a transaction:

  * ``autocommit`` - The autocommit flag MUST be set to false.
  * ``txnNumber`` - MUST be set.

* If ``runCommand`` is the first operation of the transaction:

  * ``startTransaction`` - MUST be set to true.
  * ``readConcern`` - MUST be set to the Transaction's readConcern setting if it is NOT the default.

> * See Transaction's section on `Generic RunCommand helper within a transaction <https://github.com/mongodb/specifications/blob/DRIVERS-2533-runCursorCommand/source/transactions/transactions.rst#generic-runcommand-helper-within-a-transaction>`_

Author: fixed

ReadConcern and WriteConcern
""""""""""""""""""""""""""""

> A RunCommand API MUST NOT support either Read Concern nor Write Concern options.

Author: fixed

> Additionally, unless within a transaction the command sent MUST not have any Read Concern nor Write Concern fields applied that may be inherited from client, database, or collection options.

Author: fixed

* See Read Concern's section on `Generic Command Method <https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#generic-command-method>`_
* See Write Concern's section on `Generic Command Method <https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#generic-command-method-1>`_

Retryability
""""""""""""

All commands passed to runCommand are considered non-retryable reads.

> Drivers MUST NOT inspect the command to determine if it is a write and MUST NOT attach a transaction ID.

Author: addressed

* See Retryable Reads' section on `Unsupported Read Operations <https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#unsupported-read-operations>`_
* See Retryable Writes' section on `Behavioral Changes for Write Commands <https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#behavioral-changes-for-write-commands>`_

Stable API
""""""""""

The command sent MUST attach stable API fields as configured on the MongoClient.

* See Stable API's section on `Generic Command Helper Behaviour <https://github.com/mongodb/specifications/blob/master/source/versioned-api/versioned-api.rst#generic-command-helper>`_

Client Side Operations Timeout
""""""""""""""""""""""""""""""

A driver's runCommand API MUST provide an optional timeoutMS option to support client side operations timeout.
Drivers MUST NOT attempt to check the command document for the presence of a ``maxTimeMS`` field.

> Drivers MUST document the behavior of runCommand if a ``maxTimeMS`` field  is already set on the command (ex. overwrites the field or not).

Reviewer: ```suggestion
Drivers MUST document the behavior of runCommand if a ``maxTimeMS`` field  is already set on the command (such as overwriting the command field).
```

Author: fixed.

META
====

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
document are to be interpreted as described in `RFC 2119
<https://www.ietf.org/rfc/rfc2119.txt>`_.

Specification
=============

-----
Terms
-----
Command
    A structure representing a BSON document that has a shape matching a supported MongoDB operation.

---------------------------
Implementation requirements
---------------------------

All drivers MAY offer the operations defined in the following sections.
This does not preclude a driver from offering more.

--------------
``runCommand``
--------------

The following represents how a runCommand API SHOULD be exposed.

.. code:: typescript

    interface Db {
      /**
       * Takes an argument representing an arbitrary BSON document and executes it against the server.
       */
      runCommand(command: BSONDocument, options: RunCommandOptions): BSONDocument;
    }

    interface RunCommandOptions {
      /**
       * An optional readPreference setting to apply to server selection logic.

>        * Also the value that MUST be applied to the command document as the $readPreference global command argument if not set to primary.

Reviewer: ```suggestion
       * This value MUST be applied to the command document as the $readPreference global command argument if not set to primary.
```

Author: fixed

      /**
       * An optional explicit client session.
       * The associated logical session id (`lsid`) MUST be applied to the command.
       *
       * @see https://www.mongodb.com/docs/manual/core/read-isolation-consistency-recency/#std-label-sessions
       */
      session?: ClientSession;

      /**
       * An optional duration the command should be permitted to run for.
       *
       * @ see https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst
       */
      timeoutMS?: number;
    }

Deviations
^^^^^^^^^^

* Drivers MAY rename their API to match the language syntax.

RunCommand implementation details
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

RunCommand provides a way to access MongoDB server commands directly without requiring a driver to implement a bespoke helper.
The API is intended to take a document from a user and apply a number of common driver internal concerns before forwarding the command to a server.
In general, a driver SHOULD avoid inspecting or modifying the user's command document in an effort to remain forward compatible with any potential server commands.

> Drivers that have historically modified user input SHOULD strive to instead clone the input such that appended fields do not effect the user's input.

Reviewer: ```suggestion
Drivers that have historically modified user input SHOULD strive to instead clone the input such that appended fields do not affect the user's input.
```

Author: fixed

"SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this
document are to be interpreted as described in `RFC 2119
<https://www.ietf.org/rfc/rfc2119.txt>`_.

Specification
=============

-----
Terms
-----
Command
    A structure representing a BSON document that has a shape matching a supported MongoDB operation.

---------------------------
Implementation requirements
---------------------------

All drivers MAY offer the operations defined in the following sections.
This does not preclude a driver from offering more.

--------------
``runCommand``
--------------

The following represents how a runCommand API SHOULD be exposed.

.. code:: typescript

    interface Db {
      /**
       * Takes an argument representing an arbitrary BSON document and executes it against the server.
       */
      runCommand(command: BSONDocument, options: RunCommandOptions): BSONDocument;
    }

    interface RunCommandOptions {
      /**
       * An optional readPreference setting to apply to server selection logic.
       * This value MUST be applied to the command document as the $readPreference global command argument if not set to primary.
       *

>        * @defaultValue primary

Reviewer 2: ```suggestion
       * @defaultValue ReadPreference(mode: primary)
```

This may be pedantic, but this is just the mode option. I think expressing the entire [ReadPreference structure](https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-preference) would be more clear.

I don't recall offhand if the `$readPrefrence` [global command argument]9https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#command-arguments-as-payload) takes all of the same options as the structure defined in the Server Selection spec, but the example does show it expressed as an object with a `mode` property, so I think that would still better match the suggested replacement here.

Author: The node driver, at least, will send mode, tags, maxStalenessSeconds, and hedge

       *

>        * @see https://www.mongodb.com/docs/manual/core/read-preference/

Reviewer 2: Is https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#read-preference a better reference link? I think our own specs are more verbose than the server manual.

Author: updated!

       */
      readPreference?: ReadPreference;

      /**
       * An optional explicit client session.

>        * The associated logical session id (`lsid`) MUST be applied to the command.

Author: Updated to mention 'driver'

       *

>        * @see https://www.mongodb.com/docs/manual/core/read-isolation-consistency-recency/#std-label-sessions

Reviewer 2: Likewise, should this refer to our own session spec?

Author: updated!

       */
      session?: ClientSession;

      /**
       * An optional timeout option to govern the amount of time that a single operation can execute before control is returned to the user.
       * This timeout applies to all of the work done to execute the operation, including but not limited to server selection, connection checkout, and server-side execution.
       *
       * @ see https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst
       */
      timeoutMS?: number;
    }

Deviations
^^^^^^^^^^

> * Drivers MAY rename their API to match the language syntax.

Reviewer 2: You can optionally refer to the deviations guidelines in the CRUD spec, since that's much more fleshed out. Feel free to disregard, though.

Separate note: "Deviations" is typically a top-level field for the spec itself, so I think this can be moved up and out of the API section.

Author: Moved deviations up, and linked to the CRUD spec I think it has more detail worth referring to

RunCommand implementation details
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

RunCommand provides a way to access MongoDB server commands directly without requiring a driver to implement a bespoke helper.
The API is intended to take a document from a user and apply a number of common driver internal concerns before forwarding the command to a server.
In general, a driver SHOULD avoid inspecting or modifying the user's command document in an effort to remain forward compatible with any potential server commands.

> Drivers that have historically modified user input SHOULD strive to instead clone the input such that appended fields do not affect the user's input.

Author: Updated the language here to be more strict and discuss a bit of the impacts BSON implementation has here.

OP_MSG
""""""

The ``$db`` global command argument MUST be sent on the command sent to the server and it MUST correspond to the database name the runCommand API was invoked on.

* See OP_MSG's section on `Global Command Arguments <https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#global-command-arguments>`_

ReadPreference
""""""""""""""

> A driver's runCommand implementation MUST assume all commands are read operations.

Reviewer 2: Is "MUST assume all commands are read operations" just so we don't have a logical conflict with the operation accepting a ReadPreference option?

This makes it sound as if a user can't use this to execute write commands, although I see no reason to prohibit that (they'll just have to make sure a primary is selected).

Author: I think two other components of this assumption is for performing retryability, and applying writeConcern settings both of which are respectively prohibited. So it is perhaps redundant to "assume commands are reads". For the sake of this section maybe it is important to include mention of server selection:

> For the purposes of server selection a RunCommand MUST assume all commands are read operations.

> To facilitate selecting the server the command should be sent to a driver's runCommand MUST accept an optional Read Preference option.

Reviewer 2: ```suggestion
To facilitate server selection the runCommand operation MUST accept an optional Read Preference option.
```

I'm not sure if this suggestion conveys the original intention, but the existing text definitely needs some correction.

* See Server Selection's section on `Use of read preferences with commands <https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#use-of-read-preferences-with-commands>`_

> If the provided Read Preference is NOT primary, the command sent MUST include the ``$readPreference`` global command argument.

Reviewer 2: ```suggestion
If the provided ReadPreference is NOT ``{mode: primary}``, the command sent MUST include the ``$readPreference`` global command argument.
```

Small changes. Referring to the type itself here, and also denoting the entire structure (related to my earlier suggestion regarding default value).

* See OP_MSG's section on `Global Command Arguments <https://github.com/mongodb/specifications/blob/master/source/message/OP_MSG.rst#global-command-arguments>`_

Driver Sessions
"""""""""""""""

A driver's runCommand API MUST provide an optional session option to support explicit sessions and transactions.

> If a session is not provided the driver MUST attach an implicit session if the selected server supports sessions.

Reviewer 2: Is this still relevant if drivers now assume all servers support sessions? IIRC, min supported server version is now 3.6. I think this came up in another spec, but I could be mistaken.

Author: Mongocryptd does not support sessions, the recent change to the spec related to this was that drivers no longer perform server selection to determine session support, but support for sessions is still checked before attaching an lsid

> Every ClientSession has a corresponding Logical Session ID representing the server side session ID.

Reviewer 2: ```suggestion
Every ClientSession has a corresponding Logical Session ID representing the server-side session ID.
```

The Logical Session ID MUST be included under ``lsid`` in the command sent to the server without modifying user input.

* See Driver Sessions' section on `Sending the session ID to the server on all commands <https://github.com/mongodb/specifications/blob/master/source/sessions/driver-sessions.rst#sending-the-session-id-to-the-server-on-all-commands>`_

Transactions
""""""""""""

> If RunCommand is used within a Transaction the read preference MUST be sourced from the Transaction's options.

Reviewer 2: ```suggestion
If runCommand is used within a transaction the read preference MUST be sourced from the transaction's options.
```

"runCommand" seemed consistent with prior text in the spec. If you prefer "RunCommand", I'd suggest going through and changing everything to a single form.

Transaction isn't a proper noun, so we can lowercase that.

"read preference" is fine as-is since you're not referring to the special type.

Alternatively, "transaction's options" can refer to the special type: [TransactionOptions](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#transactionoptions).

The command sent to the server MUST include the transaction specific fields, summarized as follows:

* If ``runCommand`` is executing within a transaction:

  * ``autocommit`` - The autocommit flag MUST be set to false.
  * ``txnNumber`` - MUST be set.

* If ``runCommand`` is the first operation of the transaction:

  * ``startTransaction`` - MUST be set to true.

>   * ``readConcern`` - MUST be set to the Transaction's readConcern setting if it is NOT the default.

Reviewer 2: ```suggestion
  * ``readConcern`` - MUST be set to the transaction's read concern if it is NOT the default.
```

> * See Transactions' section on `Generic RunCommand helper within a transaction <https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#generic-runcommand-helper-within-a-transaction>`_

Reviewer 2: ```suggestion
* See `Generic RunCommand helper within a transaction <https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#generic-runcommand-helper-within-a-transaction>`_ in the Transactions specification.
```

I assume this is what you meant, as there's no "Transactions" section under that heading.

ReadConcern and WriteConcern
""""""""""""""""""""""""""""

> A RunCommand API MUST NOT support Read Concern and Write Concern options.
> Additionally, unless within a transaction, the command sent MUST NOT have any Read Concern or Write Concern fields applied that may be inherited from client, database, or collection options.

Reviewer 2: ```suggestion
RunCommand MUST NOT support read concern and write concern options.

Additionally, unless executing within a transaction, the RunCommand operation MUST NOT set the ``readConcern`` or ``writeConcern`` fields in the command document. For example, default values MUST NOT be inherited from client, database, or collection options.

If the user-provided command document already includes ``readConcern`` or ``writeConcern`` fields, the values MUST be left as-is.
```

Added the last paragraph to be clear. I don't think there's anything stopping a user from supplying their own options.

Change "RunCommand" casing for consistency if needed.

* See Read Concern's section on `Generic Command Method <https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#generic-command-method>`_
* See Write Concern's section on `Generic Command Method <https://github.com/mongodb/specifications/blob/master/source/read-write-concern/read-write-concern.rst#generic-command-method-1>`_

Retryability
""""""""""""

> All commands passed to runCommand are considered non-retryable reads.

Reviewer 2: Can we just say "non-retryable" here? I don't think we need to enforce that runCommand is only for reads, if the important factor is saying it's _treated_ like a read operation so we can solicit an optional read preference (and doesn't get default routing to a primary like explicit write operations).

Author: Yea, this keeps it succinct

Drivers MUST NOT inspect the command to determine if it is a write and MUST NOT attach a ``txnNumber``.

* See Retryable Reads' section on `Unsupported Read Operations <https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#unsupported-read-operations>`_
* See Retryable Writes' section on `Behavioral Changes for Write Commands <https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#behavioral-changes-for-write-commands>`_

Stable API
""""""""""

The command sent MUST attach stable API fields as configured on the MongoClient.

* See Stable API's section on `Generic Command Helper Behaviour <https://github.com/mongodb/specifications/blob/master/source/versioned-api/versioned-api.rst#generic-command-helper>`_

Client Side Operations Timeout
""""""""""""""""""""""""""""""

> A driver's runCommand API MUST provide an optional timeoutMS option to support client side operations timeout.

Reviewer 2: ```suggestion
A driver's runCommand API MUST provide an optional ``timeoutMS`` option to support client side operations timeout.
```

If we're referring to actual field names, I think using code formatting is good for consistency.

> Drivers MUST NOT attempt to check the command document for the presence of a ``maxTimeMS`` field.

Reviewer 2: This language might be something we can use earlier when talking about not modifying existing ``readConcern`` or ``writeConcern`` fields on the command document.

Likewise, should we add similar text for the section talking about ``lsid`` setting? IIRC, drivers are supposed to append the field regardless of whether it exists on the user-supplied document or not. And depending on the driver and BSON representation I think that could result in duplicate fields or overwriting the existing value -- which we don't really care about since the user is probably doing something they shouldn't.

Author: Added notes to session, RW concern. And improved the detail in the implementation details section about modification to highlight append vs overwrite

Stable API
""""""""""

The command sent MUST attach stable API fields as configured on the MongoClient.

* See Stable API's section on `Generic Command Helper Behaviour <https://github.com/mongodb/specifications/blob/master/source/versioned-api/versioned-api.rst#generic-command-helper>`_

Client Side Operations Timeout
""""""""""""""""""""""""""""""

RunCommand MUST provide an optional ``timeoutMS`` option to support client side operations timeout.
Drivers MUST NOT attempt to check the command document for the presence of a ``maxTimeMS`` field.
Drivers MUST document the behavior of RunCommand if a ``maxTimeMS`` field  is already set on the command (such as overwriting the command field).

* See Client Side Operations Timeout's section on `runCommand <https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst#runcommand>`_
* See Client Side Operations Timeout's section on `runCommand behavior <https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst#runcommand-behavior>`_

Changelog
=========

> :2023-03-28: Add run command specification.

Author: bumped this, and I'll keep in mind to update it before we merge
